### PR TITLE
Modernize GPU particle comm

### PR DESCRIPTION
**Pull request by fweik**
Fixes #

Description of changes:
 - Replaced manual memory management by vectors
 - Two special allocators to make this possible
 - Buffer size is adapted on-demand, so the state tracking
   code could be simplified
 - General cleanup, removal of unused stuff

### CODE REVIEW

File: src/core/CudaDeviceAllocator.hpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  CudaDeviceAllocator() noexcept = default;
+  template <class U>
+  explicit CudaDeviceAllocator(const CudaDeviceAllocator<U> &) {}
+  template <class U> bool operator==(const CudaDeviceAllocator<U> &) const {
+    return true;
+  }
+  template <class U> bool operator!=(const CudaDeviceAllocator<U> &) const {
+    return false;
+  }
```
Comment (KaiSzuttor) [inline src/core/CudaDeviceAllocator.hpp:41]: can you please explain why this is necessary?

File: src/core/cuda_common_cuda.cu [hunk 2]
```diff
@@ -98,145 +83,25 @@ void _cuda_check_errors(const dim3 &block, const dim3 &grid,
...
-  if (part_index < global_part_vars_device->number_of_particles) {
-    particle_forces_device[3 * part_index + 0] = 0.0f;
-    particle_forces_device[3 * part_index + 1] = 0.0f;
-    particle_forces_device[3 * part_index + 2] = 0.0f;
-#ifdef ROTATION
-    particle_torques_device[3 * part_index + 0] = 0.0f;
-    particle_torques_device[3 * part_index + 1] = 0.0f;
-    particle_torques_device[3 * part_index + 2] = 0.0f;
-#endif
+template <class T> void resize_or_replace(device_vector<T> &vec, size_t n) {
```
Comment (KaiSzuttor) [inline src/core/cuda_common_cuda.cu:101]: please add a docstring

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]
LGTM besides some missing doc

File: src/core/cuda_common_cuda.cu (hunk 2 above)
Comment (fweik) [reply src/core/cuda_common_cuda.cu:101]: The docstring would be very similar to the function signature, are you sure
it is required for an internal function?

File: src/core/CudaDeviceAllocator.hpp (hunk 1 above)
Comment (fweik) [reply src/core/CudaDeviceAllocator.hpp:41]: Otherwise it's not an `Allocator`: https://en.cppreference.com/w/cpp/named_req/Allocator

Comment (KaiSzuttor) [reply src/core/CudaDeviceAllocator.hpp:41]: > returns true only if the storage allocated by the allocator a1 can be deallocated through a2.

so this is always fulfilled?

Comment (fweik) [reply src/core/CudaDeviceAllocator.hpp:41]: Yes, because they all call `cudaFree`.

File: src/core/cuda_common_cuda.cu (hunk 2 above)
Comment (KaiSzuttor) [reply src/core/cuda_common_cuda.cu:101]: I had the need of existence for this function in mind which should be documented as you have done by now.

**Review by KaiSzuttor — APPROVED** [review]

**Pull request merged**